packages/realm/src/schema/normalize.ts

```diff
@@ -43,163 +51,400 @@ export const PRIMITIVE_TYPES = new Set<PropertyTypeName>([
   "uuid",
 ]);
 
-function isPrimitive(type: string | undefined) {
-  return PRIMITIVE_TYPES.has(type as PropertyTypeName);
-}
+const COLLECTION_TYPES = new Set<CollectionPropertyTypeName>(["list", "dictionary", "set"]);
 
-function validateCanonicalPropertySchema(
-  objectSchemaName: string,
-  { name, type, objectType, optional }: CanonicalObjectSchemaProperty,
-) {
-  if (type === "list" && objectType === "list") {
-    throw new Error(`List property '${objectSchemaName}#${name}' cannot have list elements`);
-  }
-  if (type === "list" && !isPrimitive(objectType) && optional) {
-    throw new Error(`List property '${objectSchemaName}#${name}' of '${objectType}' elements, cannot be optional`);
-  }
-  if (objectType === "") {
-    throw new Error(`Property '${objectSchemaName}#${name}' cannot have an empty object type`);
-  }
+const COLLECTION_SHORTHAND_TO_NAME: Readonly<Record<string, string>> = {
+  "[]": "list",
+  "{}": "dictionary",
+  "<>": "set",
+};
+
+function isPrimitive(type: string | undefined): boolean {
+  return PRIMITIVE_TYPES.has(type as PrimitivePropertyTypeName);
 }
 
-export const COLLECTION_TYPES: PropertyTypeName[] = ["set", "dictionary", "list"];
+function isCollection(type: string | undefined): boolean {
+  return COLLECTION_TYPES.has(type as CollectionPropertyTypeName);
+}
 
-function removeUndefinedValues<T extends Record<string, unknown>>(obj: T): T {
-  return Object.fromEntries(Object.entries(obj).filter(([, v]) => typeof v !== "undefined")) as T;
+function isUserDefined(type: string | undefined): boolean {
+  return !!type && !(isPrimitive(type) || isCollection(type) || type === "object" || type === "linkingObjects");
```

Comment:
Yeah I think it would increase readability a bit. I changed it from the negated values to the current one as a slight performance benefit of `||` over `&&` if most property types are primitives or collections.